pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go

```diff
@@ -0,0 +1,473 @@
+								MountPath: volumeMountPathConfig,
+							},
+						},
+					},
+				},
+				DNSPolicy:                     corev1.DNSClusterFirst,
+				RestartPolicy:                 corev1.RestartPolicyAlways,
+				SchedulerName:                 corev1.DefaultSchedulerName,
+				TerminationGracePeriodSeconds: pointer.Int64Ptr(30),
+				Volumes: []corev1.Volume{
+					{
+						Name: k.secrets.Kubeconfig.Name,
+						VolumeSource: corev1.VolumeSource{
+							Secret: &corev1.SecretVolumeSource{
+								SecretName: k.secrets.Kubeconfig.Name,
+							},
+						},
+					},
+					{
+						Name: k.secrets.Server.Name,
+						VolumeSource: corev1.VolumeSource{
+							Secret: &corev1.SecretVolumeSource{
+								SecretName: k.secrets.Server.Name,
+							},
+						},
+					},
+					{
+						Name: configMap.Name,
+						VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{
+								LocalObjectReference: corev1.LocalObjectReference{
+									Name: configMap.Name,
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k *kubeScheduler) SetSecrets(secrets Secrets)          { k.secrets = secrets }
+func (k *kubeScheduler) Destroy(_ context.Context) error     { return nil }
+func (k *kubeScheduler) Wait(_ context.Context) error        { return nil }
+func (k *kubeScheduler) WaitCleanup(_ context.Context) error { return nil }
+
+func (k *kubeScheduler) emptyConfigMap() *corev1.ConfigMap {
+	return &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "kube-scheduler-config", Namespace: k.namespace}}
+}
+
+func (k *kubeScheduler) emptyVPA() *autoscalingv1beta2.VerticalPodAutoscaler {
+	return &autoscalingv1beta2.VerticalPodAutoscaler{ObjectMeta: metav1.ObjectMeta{Name: "kube-scheduler-vpa", Namespace: k.namespace}}
+}
+
+func (k *kubeScheduler) emptyService() *corev1.Service {
+	return &corev1.Service{ObjectMeta: metav1.ObjectMeta{Name: "kube-scheduler", Namespace: k.namespace}}
+}
+
+func (k *kubeScheduler) emptyDeployment() *appsv1.Deployment {
+	return &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.DeploymentNameKubeScheduler, Namespace: k.namespace}}
+}
+
+func (k *kubeScheduler) computeServerPort() int32 {
+	if k.versionConstraints.k8sGreaterEqual113 {
+		return 10259
+	}
+	return 10251
+}
+
+func (k *kubeScheduler) computeServerURIScheme() corev1.URIScheme {
+	if k.versionConstraints.k8sGreaterEqual113 {
+		return corev1.URISchemeHTTPS
+	}
+	return corev1.URISchemeHTTP
+}
+
+func (k *kubeScheduler) computeEnvironmentVariables() []corev1.EnvVar {
+	if k.config != nil && k.config.KubeMaxPDVols != nil {
+		return []corev1.EnvVar{{
+			Name:  "KUBE_MAX_PD_VOLS",
+			Value: *k.config.KubeMaxPDVols,
+		}}
+	}
+	return nil
+}
+
+func (k *kubeScheduler) computeComponentConfig() (string, string, error) {
+	var apiVersion string
+	if k.versionConstraints.k8sGreaterEqual118 {
+		apiVersion = "kubescheduler.config.k8s.io/v1alpha2"
```

Comment:
On 1.18 the scheduler crashes with the following error:
```
: v1alpha2.KubeSchedulerConfiguration.ClientConnection: v1alpha1.ClientConnectionConfiguration.BindTimeoutSeconds: TypeMeta: APIVersion: ReadObject: found unknown field: algorithmSource, error found in #10 byte of ...|thmSource":{},"apiVe|..., bigger context ...|{"algorithmSource":{},"apiVersion":"kubescheduler.config.k8s.io/v1al|...
```

The reason seems to be, that `algorithmSource` was removed in [`kubescheduler.config.k8s.io/v1alpha2.KubeSchedulerConfiguration`](https://github.com/kubernetes/kube-scheduler/blob/da23134c2b6a206bfa461e2779b32c7c93d19708/config/v1alpha2/types.go#L40).

Is there a possibility to not marshal the fields, that we don't explicitly set? (I see a lot of fields in the config API don't have `omitempty` tags). Maybe we should go back to a go template for the config 🤔

Reply:
Yes, I think it makes sense to go back to a template for the configuration. It is already cumbersome to overwrite those API versions, and it will become more cumbersome when the structure differs between the version (e.g., for the `algorithmSource` field).